lib/crawler/coordinator.rb

```diff
@@ -447,21 +447,23 @@ def extract_links(crawl_result, crawl_depth:)
     #-----------------------------------------------------------------------------------------------
     # Outputs the results of a single URL processing to an output module configured for the crawl
     def output_crawl_result(crawl_result)
-      sink.write(crawl_result).tap do |outcome|
-        # Make sure we have an outcome of the right type (helps troubleshoot sink implementations)
-        unless outcome.is_a?(Hash)
-          error = "Expected to return an outcome object from the sink, returned #{outcome.inspect} instead"
-          raise ArgumentError, error
+      Timeout.timeout(SINK_LOCK_TIMEOUT) do
```

Comment:
Is it a library function, or built-in ruby one? 

Is this one thread-safe?

Reply:
Also - how does SINK_LOCK_TIMEOUT interact with retries for sink? Can it timeout before sink finishes retrying?

Reply:
`Timeout` is built-in ruby and it appears it wouldn't be thread-safe there 🤦🏻 sorry I never thought something built-in could be so bad, it seems everywhere recommends not to use it.

> Also - how does SINK_LOCK_TIMEOUT interact with retries for sink? Can it timeout before sink finishes retrying?

There's no interaction, it's just a flat timeout.
There are two retries that would happen in sink.

- First is acquiring the lock, which isn't really a timeout, `mutex.synchronize` just waits for the lock to be lifted.
- Second is the flush (ES request) retries. In a worst-case scenario of 4 attempts each timing out at 10 seconds, it would take around 54 seconds (10 + 12 + 14 + 18) to release the lock. If the flush fails the entire payload will be dropped so the next executor that acquires the lock won't reattempt this request, it'll add its crawl result to the now-empty queue.

If we remove `Timeout` I think this could be reworked to use a `mutex.try_lock` > `mutex.unlock` block instead of `mutex.synchronize`. If the lock can't be retried it sleeps for 1 second, up to a max of maybe 120 retries (so approx. two minutes), before throwing an error. How does that sound?

Reply:
Oh it happens :)

I think indeed using mutex is best. A question raised in my head while reading through it - what happens if Elasticsearch is overloaded and start throttling?

So you have 10 threads that are getting page content and throwing it into single sink. Sink starts to slow down, but do threads keep extracting content? What happens if your timeout it 120 seconds, but sink took 118 seconds to unlock, will all threads waiting for it be able to write into it anyway?

Reply:
@artem-shelkovnikov I've changed things a bit so `Timeout` isn't used, I thought it made more sense to go by lock acquisition attempts instead (this also means if a thread acquires a lock late and takes a long time itself, it won't be unnecessarily terminated early).

Reply:
I've taken a look and haven't found anything broken with this approach. I'm also not _super_ good with concurrent operations, so some stuff might go south regardless, this is something that's just worth testing at some point (Like ingest a large site into a super small Elasticsearch instance and observe), that can of course be done later when you feel it's the right moment!